randovania/games/factorio/generator/item_cost.py

```diff
@@ -0,0 +1,155 @@
+        ingredients = []
+
+        for ingredient in recipe["ingredients"]:
+            if isinstance(ingredient, list):
+                assert len(ingredient) == 2
+                item_name, amount = ingredient
+            else:
+                assert isinstance(ingredient, dict)
+                item_name, amount = ingredient["name"], ingredient["amount"]
+
+            nested = cost_for_item(item_name)
+            ingredients.append((nested, amount))
+            steps = max(steps, nested.steps)
+            categories |= nested.categories
+
+        material, complexity = cost_for_ingredient_list(ingredients)
+
+        return ItemCost(material, complexity, steps + 1, categories)
+
+    def cost_for_item(item_name: str) -> ItemCost:
+        if item_name in item_costs:
+            return item_costs[item_name]
+        if item_name in processing_items:
+            raise ValueError(f"Loop when looking at {item_name}")
+        processing_items.add(item_name)
+
+        costs = {}
+
+        # for recipe
+        for recipe_name in sorted(item_to_recipe[item_name]):
+            recipe = recipes_raw[recipe_name]
+            count = data_parser.count_for_result(recipe, item_name)
+            recipe_complex = cost_for_recipe(recipe)
+            costs[recipe_name] = dataclasses.replace(recipe_complex, material=recipe_complex.material / count)
+
+        if costs:
+            if len(costs) > 1:
+                cost = min(costs.values(), key=lambda it: it.complexity + it.material)
+            else:
+                cost = next(iter(costs.values()))
+        else:
+            raise ValueError(f"No recipes for {item_name}")
+
+        item_costs[item_name] = dataclasses.replace(cost, is_fluid=item_name in _k_fluids)
+        processing_items.remove(item_name)
+        return cost
+
+    item_costs["iron-ore"] = ItemCost(5.0, 1, 0, {"miner"})
+    item_costs["copper-ore"] = ItemCost(5.0, 1, 0, {"miner"})
+    item_costs["coal"] = ItemCost(5.0, 1, 0, {"miner"})
+    item_costs["stone"] = ItemCost(5.0, 1, 0, {"miner"})
+    item_costs["water"] = ItemCost(0.0, 1, 0, {"pump"}, is_fluid=True)
+    item_costs["steam"] = ItemCost(1.0, 2, 1, {"boiler"}, is_fluid=True)
+    item_costs["crude-oil"] = ItemCost(2.0, 4, 0, {"pumpjack"}, is_fluid=True)
+    item_costs["wood"] = ItemCost(math.inf, 1, 0, {"hand"})
+    item_costs["raw-fish"] = ItemCost(math.inf, 1, 0, {"hand"})
+    item_costs["uranium-ore"] = cost_for_item("sulfuric-acid").combine(5.0, 2, {"miner"})
+    item_costs["uranium-235"] = cost_for_item("uranium-238").combine(1.0, 4, {"centrifuging"})
+
+    for it in list(item_to_recipe):
+        cost_for_item(it)
+
+    return item_costs
+
+
+def item_is_fluid(item_name: str) -> bool:
+    return item_name in _k_fluids
+
+
+def print_things():
+    recipes_raw = data_parser.load_recipes_raw()
+    techs_raw = data_parser.load_techs_raw()
+    complexities = cost_calculator(recipes_raw, techs_raw)
+
+    from prettytable import PrettyTable
```

Comment:
should this be a dependency? or should this function just be removed?

Reply:
Probably removed, but it's a very helpful function for debugging this